File: apps/api/plane/app/serializers/worklog.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Third party imports
from rest_framework import serializers

# Module imports
from plane.app.serializers.base import BaseSerializer
from plane.db.models import IssueWorkLog


class IssueWorkLogSerializer(BaseSerializer):
    issue_detail = serializers.SerializerMethodField()
    project_detail = serializers.SerializerMethodField()

    class Meta:
        model = IssueWorkLog
        fields = [
            "id",
            "created_at",
            "updated_at",
            "description",
            "duration",
            "created_by",
            "updated_by",
            "project_id",
            "workspace_id",
            "logged_by",
            "issue_detail",
            "project_detail",
        ]
        read_only_fields = [
            "logged_by",
            "issue",
            "workspace",
            "project",
        ]

    def get_issue_detail(self, obj):
        issue = obj.issue
        return {
            "id": str(issue.id),
            "sequence_id": issue.sequence_id,
            "name": issue.name,
        }

    def get_project_detail(self, obj):
        project = obj.project
        return {
            "id": str(project.id),
            "name": project.name,
            "identifier": project.identifier,
        }

File: packages/types/src/worklog.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TWorklogIssue = {
  id: string | undefined;
  sequence_id: number | undefined;
  name: string | undefined;
};

export type TWorklog = {
  id: string | undefined;
  description: string | undefined;
  logged_by: string | undefined;
  /** Duration in minutes */
  duration: number | undefined;
  workspace_id: string | undefined;
  project_id: string | undefined;
  issue_detail: TWorklogIssue | undefined;
  created_by: string | undefined;
  updated_by: string | undefined;
  created_at: string | undefined;
  updated_at: string | undefined;
};

export type TWorklogIssueTotalCount = {
  total_worklog: number | undefined;
};

export type TWorklogPaginatedInfo = {
  next_cursor: string | undefined;
  prev_cursor: string | undefined;
  next_page_results: boolean | undefined;
  prev_page_results: boolean | undefined;
  total_pages: number | undefined;
  count: number | undefined;
  total_count: number | undefined;
  results: TWorklog[] | undefined;
};
